python/ray/data/dataset.py

```diff
@@ -747,16 +758,39 @@ def add_column(
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
 
-        def add_column(batch: "pandas.DataFrame") -> "pandas.DataFrame":
-            batch.loc[:, col] = fn(batch)
-            return batch
+        def add_column(
+            batch: DataBatch,
+        ) -> Union["pyarrow.Array", "pandas.Series", Dict[str, "np.ndarray"]]:
+            column = fn(batch)
+            if batch_format == "pandas":
+                batch.loc[:, col] = column
```

Comment:
Let's assert that we get the column of appropriate type

Reply:
Ditto for all other formats

Reply:
Mhhh no, these changes are up!